File: DEC_SIM_GA.py

"""
Keras implementation for Deep Embedded Clustering (DEC) algorithm:
        Junyuan Xie, Ross Girshick, and Ali Farhadi. Unsupervised deep embedding for clustering analysis. ICML 2016.
Usage:
    use `python DEC.py -h` for help.
Author:
    Xifeng Guo. 2017.1.30
"""
from time import time
import numpy as np
from tensorflow.keras import backend as K
from tensorflow.python.keras.layers import Layer, InputSpec
from tensorflow.keras.layers import Dense, Input, Dropout, BatchNormalization
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import SGD
from tensorflow.keras import callbacks
from tensorflow.keras.initializers import VarianceScaling
from sklearn.cluster import KMeans
#import metrics
from sklearn import metrics 
import matplotlib.pyplot as plt
import tensorflow as tf
from tensorflow import keras



def autoencoder(dims, act='tanh', init='glorot_uniform'):
    """
    Fully connected auto-encoder model, symmetric.
    Arguments:
        dims: list of number of units in each layer of encoder. dims[0] is input dim, dims[-1] is units in hidden layer.
            The decoder is symmetric with encoder. So number of layers of the auto-encoder is 2*len(dims)-1
        act: activation, not applied to Input, Hidden and Output layers
    return:
        (ae_model, encoder_model), Model of autoencoder and model of encoder
    """
    n_stacks = len(dims) - 1
    # input
    x = Input(shape=(dims[0],), name='input')
    h = x

    # internal layers in encoder
    for i in range(n_stacks-1):
        h = Dense(dims[i + 1], activation=act, kernel_initializer=init, name='encoder_%d' % i)(h)
        #h = BatchNormalization()(h)
        h = Dropout(0.4)(h)
    # hidden layer
    h = Dense(dims[-1], kernel_initializer=init, name='encoder_%d' % (n_stacks - 1))(h)  # hidden layer, features are extracted from here

    y = h
    # internal layers in decoder
    for i in range(n_stacks-1, 0, -1):
        y = Dense(dims[i], activation=act, kernel_initializer=init, name='decoder_%d' % i)(y)
        #y = BatchNormalization()(y)
        y = Dropout(0.4)(y)

    # output
    y = Dense(dims[0], kernel_initializer=init, name='decoder_0')(y)

    return Model(inputs=x, outputs=y, name='AE'), Model(inputs=x, outputs=h, name='encoder')

 # ------------------------------------------------------------------------------------


class ClusteringLayer(Layer):
    """
    Clustering layer converts input sample (feature) to soft label, i.e. a vector that represents the probability of the
    sample belonging to each cluster. The probability is calculated with student's t-distribution.
    # Example
    ```
        model.add(ClusteringLayer(n_clusters=10))
    ```
    # Arguments
        n_clusters: number of clusters.
        weights: list of Numpy array with shape `(n_clusters, n_features)` witch represents the initial cluster centers.
        alpha: parameter in Student's t-distribution. Default to 1.0.
    # Input shape
        2D tensor with shape: `(n_samples, n_features)`.
    # Output shape
        2D tensor with shape: `(n_samples, n_clusters)`.
    """

    def __init__(self, n_clusters, weights=None, alpha=1.0, **kwargs):
        if 'input_shape' not in kwargs and 'input_dim' in kwargs:
            kwargs['input_shape'] = (kwargs.pop('input_dim'),)
        super(ClusteringLayer, self).__init__(**kwargs)
        self.n_clusters = n_clusters
        self.alpha = alpha
        self.initial_weights = weights
        self.input_spec = InputSpec(ndim=2)

    def build(self, input_shape):
        assert len(input_shape) == 2
        input_dim = input_shape[1]
        self.input_spec = InputSpec(dtype=K.floatx(), shape=(None, input_dim))
        self.clusters = self.add_weight(shape=(self.n_clusters, input_dim), initializer='glorot_uniform', name='clusters')
        if self.initial_weights is not None:
            self.set_weights(self.initial_weights)
            del self.initial_weights
        self.built = True

    def call(self, inputs, **kwargs):
        """ student t-distribution, as same as used in t-SNE algorithm.
                 q_ij = 1/(1+dist(x_i, u_j)^2), then normalize it.
        Arguments:
            inputs: the variable containing data, shape=(n_samples, n_features)
        Return:
            q: student's t-distribution, or soft labels for each sample. shape=(n_samples, n_clusters)
        """
        q = 1.0 / (1.0 + (K.sum(K.square(K.expand_dims(inputs, axis=1) - self.clusters), axis=2) / self.alpha))
        q **= ((self.alpha + 1.0) / 2.0)
        q = K.transpose(K.transpose(q) / K.sum(q, axis=1))
        return q
    
   

    def compute_output_shape(self, input_shape):
        assert input_shape and len(input_shape) == 2
        return input_shape[0], self.n_clusters

    def get_config(self):
        config = {'n_clusters': self.n_clusters}
        base_config = super(ClusteringLayer, self).get_config()
        return dict(list(base_config.items()) + list(config.items()))

# ------------------------------------------------------------------------------------
class DEC(object):
    def __init__(self,
                 dims,
                 n_clusters=5,
                 alpha=1.0,
                 init='glorot_uniform'):

        super(DEC, self).__init__()

        self.dims = dims
        self.input_dim = dims[0]
        self.n_stacks = len(self.dims) - 1

        self.n_clusters = n_clusters
        self.alpha = alpha
        self.autoencoder, self.encoder = autoencoder(self.dims, init=init)

        # prepare DEC model
        clustering_layer = ClusteringLayer(self.n_clusters, name='clustering')(self.encoder.output)
        self.model = Model(inputs=self.encoder.input, outputs=clustering_layer)

    def pretrain(self, x, y=None, optimizer='rmsprop', epochs=100, batch_size=200, save_dir='results/temp'):
        print('...Pretraining...')
        self.autoencoder.compile(optimizer=optimizer, loss='mse', metrics='accuracy')

        csv_logger = callbacks.CSVLogger(save_dir + '/pretrain_log.csv')
        cb = [csv_logger]
    
        # begin pretraining
        t0 = time()
        for i in range(0, 1):            
            self.autoencoder.fit(x, x, batch_size=batch_size, epochs=epochs, callbacks=cb)
            print('Pretraining time: %ds' % round(time() - t0))
        # self.autoencoder.save_weights(save_dir + '/ae_weights.h5')
        # print('Pretrained weights are saved to %s/ae_weights.h5' % save_dir)
            self.pretrained = True

    def load_weights(self, weights):  # load weights of DEC model
        self.model.load_weights(weights)

    def extract_features(self, x):
        return self.encoder.predict(x)

    def predict(self, x):  # predict cluster labels using the output of clustering layer
        q = self.model.predict(x, verbose=0)
        return q.argmax(1)

    @staticmethod
    def target_distribution(q):
        weight = q ** 2 / q.sum(0)
        return (weight.T / weight.sum(1)).T

    def compile(self, optimizer='sgd', loss='kld'):
        self.model.compile(optimizer=optimizer, loss=loss)

    def fit(self, x, y=None, maxiter=2e4, batch_size=256, tol=1e-3,
            update_interval=140, save_dir='./results/temp'):

        print('Update interval', update_interval)
        save_interval = int(x.shape[0] / batch_size) * 5  # 5 epochs
        print('Save interval', save_interval)

        # Step 1: initialize cluster centers using k-means
        t1 = time()
        print('Initializing cluster centers with k-means.')
        kmeans = KMeans(n_clusters=self.n_clusters, n_init=20)
        y_pred = kmeans.fit_predict(self.encoder.predict(x))
        y_pred_last = np.copy(y_pred)
        self.model.get_layer(name='clustering').set_weights([kmeans.cluster_centers_])
        # Step 2: deep clustering
        # logging file
        import csv
        logfile = open(save_dir + '/dec_log.csv', 'w')
        logwriter = csv.DictWriter(logfile, fieldnames=['iter', 'acc', 'nmi', 'ari', 'loss'])
        logwriter.writeheader()

        loss = 1000
        index = 0
        index_array = np.arange(x.shape[0])
        for ite in range(int(maxiter)):
            if ite % update_interval == 0:
                q = self.model.predict(x, verbose=0)
                p = self.target_distribution(q)  # update the auxiliary target distribution p
                # evaluate the clustering performance
                y_pred = q.argmax(1)
                          
                # plot
                plt.scatter(np.arange(0,len(y_pred)),y_pred)
                plt.show()
                
                if y is not None:
                    acc = np.round(metrics.accuracy_score(y_pred_last, y_pred), 5)
                    nmi = np.round(metrics.normalized_mutual_info_score(y_pred_last, y_pred), 5)
                    ari = np.round( metrics.adjusted_mutual_info_score(y_pred_last, y_pred), 5)
                    loss = np.round(loss, 5)
                    logdict = dict(iter=ite, acc=acc, nmi=nmi, ari=ari, loss=loss)
                    logwriter.writerow(logdict)
                    
                    print('Iter %d: acc = %.5f, nmi = %.5f, ari = %.5f' % (ite, acc, nmi, ari), ' ; loss=', loss)

                # check stop criterion
                delta_label = np.sum(y_pred != y_pred_last).astype(np.float32) / y_pred.shape[0]
                y_pred_last = np.copy(y_pred)
                if ite > 0 and  delta_label < tol:
                    print('delta_label ', delta_label, '< tol ', tol)
                    print('Reached tolerance threshold. Stopping training.')
                    logfile.close()
                    break

            # train on batch
            # if index == 0:
            #     np.random.shuffle(index_array)
            idx = index_array[index * batch_size: min((index+1) * batch_size, x.shape[0])]
            loss = self.model.train_on_batch(x=x[idx], y=p[idx])
            index = index + 1 if (index + 1) * batch_size <= x.shape[0] else 0

            # save intermediate model
            #if ite % save_interval == 0:
                #print('saving model to:', save_dir + '/DEC_model_' + str(ite) + '.h5')
                #self.model.save_weights(save_dir + '/DEC_model_' + str(ite) + '.h5')

            ite += 1

        # save the trained model
        #logfile.close()
        #print('saving model to:', save_dir + '/DEC_model_final.h5')
        #self.model.save_weights(save_dir + '/DEC_model_final.h5')

        return y_pred

# --------------------------------------------------------------------------------------

if __name__ == "__main__":
    # setting the hyper parameters
    import argparse

    parser = argparse.ArgumentParser(description='train',
                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    # parser.add_argument('--dataset', default='mnist',
    #                     choices=['mnist', 'fmnist', 'usps', 'reuters10k', 'stl'])
    parser.add_argument('--batch_size', default=256, type=int)
    parser.add_argument('--maxiter', default=2e4, type=int)
    parser.add_argument('--pretrain_epochs', default=None, type=int)
    parser.add_argument('--update_interval', default=None, type=int)
    parser.add_argument('--tol', default=0.001, type=float)
    parser.add_argument('--ae_weights', default=None)
    parser.add_argument('--save_dir', default='results')
    args = parser.parse_args()
    print(args)
    import os
    if not os.path.exists(args.save_dir):
        os.makedirs(args.save_dir)
        
        
        
    # load dataset ------------------------------------------------------------

    import scipy.io as sio
    
    # x_data= sio.loadmat('D:/My works/Current/Deep clustering/DC_Simulated data/inDaGA_M1.mat')
    # y_label=sio.loadmat('D:/My works/Current/Deep clustering/DC_Simulated data/CC_M1_idx.mat')
    # indata=x_data['inDaGA_M1']
    # x=indata/255
    # y=np.concatenate(y_label['Clu_idx2'])
    
    
    # x_data= sio.loadmat('D:/My works/Current/S_ConClust/ERP_CC_GA_Toolbox/full/inDaGA.mat')
    # y_label=sio.loadmat('D:/My works/Current/S_ConClust/ERP_CC_GA_Toolbox/full/CC_label.mat')
    # indata=x_data['inDaGA_M1']
    # x=indata/255
    # y=np.concatenate(y_label['CC_label'])
    
    x_data= sio.loadmat('C:/Users/Rza/Google Drive/Current/Deep clustering/SIM_GA_DC_NL/SimData_Gans.mat')
    y_label=sio.loadmat('C:/Users/Rza/Google Drive/Current/Deep clustering/SIM_GA_DC_NL/CC_GT.mat')
    
    dataAll=x_data['SimData_Gans']
    x=dataAll[:,:,1]/255.0 # put them in "n-1" starts from 0 :p
    y=np.concatenate(y_label['CC_GT'])
    
           
    
    plt.scatter(np.arange(0,len(y)),y)
    plt.show()
                
    n_clusters = len(np.unique(y)) 
    init = 'glorot_uniform'
    pretrain_optimizer = 'rmsprop' #rmsprop
    update_interval = 20
    pretrain_epochs = 50 
    init = VarianceScaling(scale=1. / 3., mode='fan_in',
                               distribution='uniform')  # [-limit, limit], limit=sqrt(1./fan_in)
    pretrain_optimizer = SGD(lr=1, momentum=0.9)
    
    if args.update_interval is not None:
        update_interval = args.update_interval
    if args.pretrain_epochs is not None:
        pretrain_epochs = args.pretrain_epochs

    # prepare the DEC model
    dec = DEC(dims=[x.shape[-1], 256, 512, 128, 6], n_clusters=n_clusters, init=init)

    dec.pretrain(x=x, y=y, optimizer=pretrain_optimizer,
                     epochs=pretrain_epochs, batch_size=args.batch_size,
                     save_dir=args.save_dir)

    dec.model.summary()
    t0 = time()
    dec.compile(optimizer=SGD(0.01, 0.9), loss='kld')
    y_pred = dec.fit(x, y=y, tol=args.tol, maxiter=args.maxiter, batch_size=args.batch_size,
                     update_interval=update_interval, save_dir=args.save_dir)
    
    sio.savemat('C:/Users/Rza/Google Drive/Current/Deep clustering/SIM_GA_DC_NL/DEC_idx.mat',{'y_pred': y_pred})
    
    
    # print('acc:', metrics.normalized_mutual_info_score(y, y_pred))
    # print('clustering time: ', (time() - t0))

File: DEC_SIM_GA.R
# Deep embedded clustering

rm(list = ls())
gc()  #free up memrory and report the memory usage.

# loding the libray --------------------------------

library(keras)
linrK <- keras::backend()
library(ggplot2)
library(R.matlab) # reading/writing .mat files
library(tidyverse)
library(tensorflow)
library(MLmetrics)
library(aricode)

#model defination --------------------------------

# defination of clustring layer
# 1- building (input_shape), 2- call, 3- comput_output(input_shape)
# Clusterlayer converts input sample (feature) to soft label.

# Example
# ```
# model.add(ClusterLayer(n_clusters=10))
# ```
# Arguments
# n_clusters: number of clusters.
# weights: list of Numpy array with shape `(n_clusters, n_features)` witch represents the initial cluster centers.
# alpha: degrees of freedom parameter in Student's t-distribution. Default to 1.0.
#     # Input shape
#         2D tensor with shape: `(n_samples, n_features)`.
#     # Output shape
#         2D tensor with shape: `(n_samples, n_clusters)`.



# loading data fro mat file  -------------------------------------------------------------------

# inData <- readMat("D:/My works/Current/Deep clustering/ERP_CORE/P3_Data.mat")
# Lab <- readMat("D:/My works/Current/Deep clustering/ERP_CORE/CC_idx.mat")
# 
# iData=as.matrix(inData$outData)
# Lb=as.matrix(Lab$CC.Label)
inData1 <- readMat("D:/My works/Current/Deep clustering/SIM_GA_DC_NL/Sim_MS3_data/SimData_Gans.mat") # data without noise
iData1=as.matrix(inData1$SimData.Gans[,,7]) # selecting dataset for training

inData <- readMat("D:/My works/Current/Deep clustering/SIM_GA_DC_NL/Sim_MS3_data/SimData_Gans.mat") # noisy data
iData=as.matrix(inData$SimData.Gans[,,6]) # selecting dataset for training
# noise levels : 1=-10dB, 2=-5dB, 3=0dB, 4=5dB, 5=10dB, 6=20dB, 7=50dB

Lab <- readMat("D:/My works/Current/Deep clustering/SIM_GA_DC_NL/CC_GT.mat") # ground truth labels
Lb=as.matrix(Lab$CC.GT) # no noise and ground-truth
# Lab <- readMat("C:/Users/Rza/Google Drive/Current/Deep clustering/SIM_GA_DC_NL/CC_label.mat")
# Lb=as.matrix(Lab$CC.label)

# Initialization ------------------------------------------------------------------------------
nbcl=6
batchSz=150L
ep=200L


# ---------------------------------------------------------------------------------------------------------------

# Fully connected auto-encoder model, symmetric.
# Arguments:
#   dims: list of number of units in each layer of encoder. dims[1] is input dim, dims[-1] is units in hidden layer.
# The decoder is symmetric with encoder. So number of layers of the auto-encoder is 2*len(dims)-1
# act: activation, not applied to Input, Hidden and Output layers
# return:
#   (ae_model, encoder_model), Model of autoencoder and model of encoder

createAutoencoderModel <- function( numberOfUnitsPerLayer,
                                    activation = 'tanh',
                                    initializer = 'glorot_uniform' )
{
  numberOfEncodingLayers <- as.integer(length( numberOfUnitsPerLayer ) - 1)
  # input of AE
  inputs <- layer_input( shape =c(as.integer(numberOfUnitsPerLayer[1])) )
  encoder <- inputs
  
  # internal layers in encoder
  for( i in seq_len( numberOfEncodingLayers - 1 ) )
  {
    encoder <- encoder %>%
      layer_dense( numberOfUnitsPerLayer[i+1],
                   activation = activation, kernel_initializer = initializer )
  }
  
  # hidden layer
  encoder <- encoder %>%
    layer_dense( units = tail( numberOfUnitsPerLayer, 1 ) )
  
  autoencoder <- encoder
  
  # internal layers in decoder
  for( i in seq( from = numberOfEncodingLayers, to = 2, by = -1 ) )
  {
    autoencoder <- autoencoder %>%
      layer_dense( numberOfUnitsPerLayer[i],
                   activation = activation, kernel_initializer = initializer )
  }
  
  # output
  autoencoder <- autoencoder %>%
    layer_dense( numberOfUnitsPerLayer[1], activation = "linear", kernel_initializer = initializer )
  
  return( list(
    autoencoderModel = keras_model( inputs = inputs, outputs = autoencoder ),
    encoderModel = keras_model( inputs = inputs, outputs = encoder ) ) )
}


# ------------------------------------------------------------------------------------------------------

# """
#     Clustering layer converts input sample (feature) to soft label, i.e. a vector that represents the probability of the
#     sample belonging to each cluster. The probability is calculated with student's t-distribution.
#     # Example
#     ```
#         model.add(ClusteringLayer(n_clusters=10))
#     ```
#     # Arguments
#         n_clusters: number of clusters.
#         weights: list of Numpy array with shape `(n_clusters, n_features)` witch represents the initial cluster centers.
#         alpha: parameter in Student's t-distribution. Default to 1.0.
#     # Input shape
#         2D tensor with shape: `(n_samples, n_features)`.
#     # Output shape
#         2D tensor with shape: `(n_samples, n_clusters)`.
#     in fact this function will return us the Q matrix (soft assignment) //me

ClusteringLayer <- R6::R6Class( "ClusteringLayer",
                                inherit = KerasLayer,
                                lock_objects = FALSE,
                                # I Changed nu...rs=10 to Null
                                public = list(numberOfClusters = NULL, initialClusterWeights = NULL,
                                              alpha = 1.0, name = '',
                                              
                                              initialize = function( numberOfClusters,
                                                                     initialClusterWeights = NULL, alpha = 1.0, name = '' )
                                              {
                                                self$numberOfClusters <- as.integer(numberOfClusters)
                                                self$initialClusterWeights <- initialClusterWeights
                                                self$alpha <- alpha
                                                self$name <- name
                                              },
                                              
                                              build = function( input_shape )
                                              {
                                                if( length( input_shape ) != 2 )
                                                {
                                                  stop( paste0( "input_shape is not of length 2." ) )
                                                }
                                                
                                                self$clusters <- self$add_weight(
                                                  shape = list( self$numberOfClusters, input_shape[[2]] ),
                                                  initializer = 'glorot_uniform', name = 'clusters' )
                                                
                                                if( ! is.null( self$initialClusterWeights ) )
                                                {
                                                  self$set_weights( self$initialClusterWeights )
                                                  self$initialClusterWeights <- NULL
                                                }
                                                self$built <- TRUE
                                              },
                                              
                                              call = function( inputs, mask = NULL )
                                              {
                                                # Uses Student t-distribution (same as t-SNE)
                                                # inputs are the variable containing the data, shape = ( numberOfSamples, numberOfFeatures )
                                                
                                                K <- keras::backend()
                                                
                                                q <- 1.0 / ( 1.0 + (K$sum( K$square(
                                                  K$expand_dims( inputs, axis = 1L ) - self$clusters ), axis = 2L ) / self$alpha)) 
                                                q <- q^( ( self$alpha + 1.0 ) / 2.0 )
                                                q <- K$transpose( K$transpose( q ) / K$sum( q, axis = 1L ) )
                                                
                                                return( q )
                                              },
                                              
                                              compute_output_shape = function( input_shape )
                                              {
                                                return( list( input_shape[[1]], self$numberOfClusters ) )
                                              }
                                )
)

# Layer Wrapper Function

layer_clustering <- function( object,
                              numberOfClusters, initialClusterWeights = NULL,
                              alpha = 1.0, name = '' )
{
  create_layer( ClusteringLayer, object,
                list( numberOfClusters = numberOfClusters,
                      initialClusterWeights = initialClusterWeights,
                      alpha = alpha, name = name )
  )
}

# ----------------------------------------------------------------------------------------------------------
#' Deep embedded clustering (DEC) model class

DeepEmbeddedClusteringModel <- R6::R6Class( "DeepEmbeddedClusteringModel",
                                            
                                            inherit = NULL,
                                            lock_objects = FALSE,
                                            
                                            public = list(
                                              numberOfUnitsPerLayer = NULL,
                                              numberOfClusters = nbcl,
                                              alpha = 1.0,
                                              initializer = 'glorot_uniform',
                                              convolutional = FALSE,
                                              inputImageSize = NULL,
                                              
                                              initialize = function( numberOfUnitsPerLayer,
                                                                     numberOfClusters, alpha = 1.0, initializer = 'glorot_uniform',
                                                                     convolutional = FALSE, inputImageSize = NULL )
                                              {
                                                self$numberOfUnitsPerLayer <- as.integer(numberOfUnitsPerLayer)
                                                self$numberOfClusters <- as.integer(numberOfClusters)
                                                self$alpha <- alpha
                                                self$initializer <- initializer
                                                self$convolutional <- convolutional
                                                self$inputImageSize <- as.integer(inputImageSize)
                     
                                                ae <- createAutoencoderModel( self$numberOfUnitsPerLayer,
                                                                              initializer = self$initializer )
                                                self$autoencoder <- ae$autoencoderModel
                                                self$encoder <- ae$encoderModel
                                             
                                                # prepare DEC model wrapping -----------------------------------------------
                                                
                                                clusteringLayer <- self$encoder$output %>%
                                                layer_clustering( self$numberOfClusters, name = "clustering" )
                                                
                                                self$model <- keras_model( inputs = self$encoder$input, outputs = clusteringLayer )
                                                },
                                              # here the model for reading input and deliver soft assignment is ready //me
                                  
                                              
                                              pretrain = function( x, y , optimizer=optimizer_rmsprop(learning_rate = 0.001), epochs = ep , batchSize = batchSz )
                                              {
                                                cat('.................... pretraining ....................... \n')
                                                
                                                self$autoencoder$compile( optimizer = optimizer, loss = 'mse' )
                                                self$autoencoder$fit( x, y, batch_size = batchSize, epochs = epochs)
                                                self$pretrained = TRUE
                                              },
                                              
                                              loadWeights = function( weights )
                                              {
                                                self$model$load_weights( weights )
                                              },
                                              
                                              extractFeatures = function( x )
                                              {
                                                return(self$encoder$predict( x ))
                                              },
                                              
                                              predictClusterLabels = function( x )
                                              {
                                                clusterProbabilities <- self$model$predict( x, verbose = 0 )
                                                return( max.col( clusterProbabilities ) )
                                              },
                                              
                                              targetDistribution = function( q )
                                              {
                                                weight <- q^2 / colSums( q )
                                                p <- t( t( weight ) / rowSums( weight ) )
                                                return( p )
                                              },
                                              
                                              compile = function( optimizer = 'sgd', loss =  'kld', lossWeights = NULL )
                                              {
                                                self$model$compile( optimizer = optimizer, loss = loss, loss_weights = lossWeights )
                                              },
                                              
                                              fit = function( x, y ,maxNumberOfIterations = 2e4, batchSize = batchSz , tolerance = 1e-3, updateInterval = 10 )
                                              {
                                                
                                                # Initialize clusters using k-means
                                                
                                                cat('Initializing cluster centers with k-means.\n')
                                                km <- stats::kmeans( self$encoder$predict( x ),
                                                                     centers = self$numberOfClusters, nstart = 20L )
                                                currentPrediction <- km$cluster #fitted( km )
                                                
                                                plot(currentPrediction, col="red")
                                                title(main = 'current prediction')  
                                                previousPrediction <- currentPrediction
                                                
                                                self$model$get_layer( name = 'clustering' )$set_weights( list( km$centers ) )
                                                
                                                # Deep clustering ----------------------------------------------------------
                                                
                                                loss <- 1000
                                                index <- 0
                                                indexArray <- 1:( dim( x )[1] )
                                                
                                                for( i in seq_len( maxNumberOfIterations ) )
                                                {
                                                  if( i %% updateInterval == 1 )
                                                  { 
                                                    q <- self$model$predict( x, verbose = 0 )
                                                    p <- self$targetDistribution( q ) # update the auxiliary target distribution p
                                          
                                                    # evaluate the clustering performance
                                                    currentPrediction <- max.col( q )
                                                    
                                                    plot(currentPrediction, col="blue")
                                                    title(main = 'current prediction')
                                                    
                                                   
                                                    # Met stopping criterion --------------------------------------------------
                                                    # evaluate the clustering performance
                                                    deltaLabel <- sum( currentPrediction != previousPrediction ) / length( currentPrediction )
                                                    
                                                    cat( "Itr", i, ": ( out of", maxNumberOfIterations,
                                                         "): loss = [", unlist( loss ), "], deltaLabel =", deltaLabel,", ACC= ", Accuracy(previousPrediction, currentPrediction),
                                                         ", NMI= ", NMI(previousPrediction, currentPrediction), "\n", sep = ' ' )
                                                    
                                                    previousPrediction <- currentPrediction
                                                    
                                                    # cat("Iteration ", i, ": (out of ", maxNumberOfIterations, "), loss = ", loss, ", deltaLabel = ", deltaLabel, "\n", sep = '' )
                                                    
                                                    if( i > 1 && deltaLabel < tolerance )
                                                    {
                                                      print('Reached tolerance threshold. Stopping training......')
                                                      break
                                                    }
                                                  }
                                                  
                                                  # train on batch
                                                  batchIndices <- indexArray[( index * batchSize + 1 ):min( ( index + 1 ) * batchSize, dim( x )[1] )]
                                                  
                                                  loss <- self$model$train_on_batch( x = x[batchIndices,], y = p[batchIndices,] )

                                                  if( ( index + 1 ) * batchSize + 1 <= dim( x )[1] )
                                                  {
                                                    index <- index + 1
                                                  } else {
                                                    index <- 0
                                                  }
                                                }
                                                return( currentPrediction )
                                              }

                                            )
)


# defining training and test sets ------------------

d_dim=ncol(iData)
x <- iData
y <-iData1  #as.vector(Lb)
numberOfClusters <- nbcl # length( unique( Lb ))
numberOfPixels <- d_dim 

initializer <- initializer_variance_scaling(
  scale = 1/3, mode = 'fan_in', distribution = 'uniform' )

pretrainOptimizer <- optimizer_sgd( learning_rate = 0.001, momentum = 0.9 )
initializer='glorot_uniform'

decModel <- DeepEmbeddedClusteringModel$new(
  numberOfUnitsPerLayer = c( numberOfPixels, 128, 256, 256, nbcl ),
  numberOfClusters = numberOfClusters, initializer = initializer )


decModel$pretrain( x=x, y=y , optimizer = pretrainOptimizer,
                   epochs = ep, batchSize = batchSz )


decModel$compile( optimizer = optimizer_sgd( learning_rate = 0.001, momentum = 0.9 ), loss = 'kld')

yPredicted <- decModel$fit( x, y, maxNumberOfIterations = 2000, batchSize = batchSz,
                            tolerance = 1e-2, updateInterval = 5 )


#Saving the labels

writeMat("D:/My works/Current/Deep clustering/SIM_GA_DC_NL/DEC_lb.mat", yPredicted=yPredicted)
